File: library/src/main/java/org/xmtp/android/library/codecs/ReactionV2Codec.kt
package org.xmtp.android.library.codecs

import uniffi.xmtpv3.FfiReactionAction
import uniffi.xmtpv3.FfiReactionPayload
import uniffi.xmtpv3.FfiReactionSchema
import uniffi.xmtpv3.decodeReaction
import uniffi.xmtpv3.encodeReaction

val ContentTypeReactionV2 =
    ContentTypeIdBuilder.builderFromAuthorityId(
        "xmtp.org",
        "reaction",
        versionMajor = 2,
        versionMinor = 0,
    )

private fun ReactionAction.toFfi(): FfiReactionAction =
    when (this) {
        ReactionAction.Added -> FfiReactionAction.ADDED
        ReactionAction.Removed -> FfiReactionAction.REMOVED
        ReactionAction.Unknown -> FfiReactionAction.UNKNOWN
    }

private fun FfiReactionAction.toReactionAction(): ReactionAction =
    when (this) {
        FfiReactionAction.ADDED -> ReactionAction.Added
        FfiReactionAction.REMOVED -> ReactionAction.Removed
        FfiReactionAction.UNKNOWN -> ReactionAction.Unknown
    }

private fun ReactionSchema.toFfi(): FfiReactionSchema =
    when (this) {
        ReactionSchema.Unicode -> FfiReactionSchema.UNICODE
        ReactionSchema.Shortcode -> FfiReactionSchema.SHORTCODE
        ReactionSchema.Custom -> FfiReactionSchema.CUSTOM
        ReactionSchema.Unknown -> FfiReactionSchema.UNKNOWN
    }

private fun FfiReactionSchema.toReactionSchema(): ReactionSchema =
    when (this) {
        FfiReactionSchema.UNICODE -> ReactionSchema.Unicode
        FfiReactionSchema.SHORTCODE -> ReactionSchema.Shortcode
        FfiReactionSchema.CUSTOM -> ReactionSchema.Custom
        FfiReactionSchema.UNKNOWN -> ReactionSchema.Unknown
    }

private fun Reaction.toFfiPayload(): FfiReactionPayload =
    FfiReactionPayload(
        reference = reference,
        referenceInboxId = referenceInboxId.orEmpty(),
        action = action.toFfi(),
        content = content,
        schema = schema.toFfi(),
    )

private fun FfiReactionPayload.toReaction(): Reaction =
    Reaction(
        reference = reference,
        action = action.toReactionAction(),
        content = content,
        schema = schema.toReactionSchema(),
        referenceInboxId = referenceInboxId.ifEmpty { null },
    )

data class ReactionV2Codec(
    override var contentType: ContentTypeId = ContentTypeReactionV2,
) : ContentCodec<Reaction> {
    override fun encode(content: Reaction): EncodedContent =
        EncodedContent.parseFrom(encodeReaction(content.toFfiPayload()))

    override fun decode(content: EncodedContent): Reaction = decodeReaction(content.toByteArray()).toReaction()

    override fun fallback(content: Reaction): String? =
        when (content.action) {
            ReactionAction.Added -> "Reacted \"${content.content}\" to an earlier message"
            ReactionAction.Removed -> "Removed \"${content.content}\" from an earlier message"
            else -> null
        }

    override fun shouldPush(content: Reaction): Boolean = content.action == ReactionAction.Added
}

File: sdks/ios/Sources/XMTPiOS/Codecs/ReactionV2Codec.swift
//
//  ReactionV2Codec.swift
//
//
//  Created by Naomi Plasterer on 7/26/23.
//

import Foundation

public let ContentTypeReactionV2 = ContentTypeID(
	authorityID: "xmtp.org",
	typeID: "reaction",
	versionMajor: 2,
	versionMinor: 0,
)

public struct ReactionV2Codec: ContentCodec {
	public typealias T = Reaction
	public var contentType = ContentTypeReactionV2

	public init() {}

	public func encode(content: Reaction) throws -> EncodedContent {
		// Convert Reaction to FfiReactionPayload for encoding
		let ffiReaction = FfiReactionPayload(
			reference: content.reference,
			referenceInboxId: content.referenceInboxId ?? "",
			action: content.action.toFfiReactionAction(),
			content: content.content,
			schema: content.schema.toFfiReactionSchema(),
		)
		return try EncodedContent(serializedBytes: encodeReaction(reaction: ffiReaction))
	}

	public func decode(content: EncodedContent) throws -> Reaction {
		let ffiReaction = try decodeReaction(bytes: content.serializedBytes())
		// Convert FfiReactionPayload to Reaction
		return Reaction(
			reference: ffiReaction.reference,
			action: ReactionAction.fromFfiReactionAction(ffiReaction.action),
			content: ffiReaction.content,
			schema: ReactionSchema.fromFfiReactionSchema(ffiReaction.schema),
			referenceInboxId: ffiReaction.referenceInboxId,
		)
	}

	public func fallback(content: Reaction) throws -> String? {
		switch content.action {
		case .added:
			"Reacted \"\(content.content)\" to an earlier message"
		case .removed:
			"Removed \"\(content.content)\" from an earlier message"
		case .unknown:
			nil
		}
	}

	public func shouldPush(content: Reaction) throws -> Bool {
		switch content.action {
		case .added:
			true
		default:
			false
		}
	}
}

// MARK: - Conversion Extensions

extension ReactionAction {
	func toFfiReactionAction() -> FfiReactionAction {
		switch self {
		case .added:
			.added
		case .removed:
			.removed
		case .unknown:
			.unknown
		}
	}

	static func fromFfiReactionAction(_ action: FfiReactionAction) -> ReactionAction {
		switch action {
		case .added:
			.added
		case .removed:
			.removed
		case .unknown:
			.unknown
		}
	}
}

extension ReactionSchema {
	func toFfiReactionSchema() -> FfiReactionSchema {
		switch self {
		case .unicode:
			.unicode
		case .shortcode:
			.shortcode
		case .custom:
			.custom
		case .unknown:
			.unknown
		}
	}

	static func fromFfiReactionSchema(_ schema: FfiReactionSchema) -> ReactionSchema {
		switch schema {
		case .unicode:
			.unicode
		case .shortcode:
			.shortcode
		case .custom:
			.custom
		case .unknown:
			.unknown
		}
	}
}
